File: Leetcode Daily Challenge/016-October-2024/020-parsing-a-boolean-expression.cpp
// https://leetcode.com/problems/parsing-a-boolean-expression/



#include <bits/stdc++.h>
using namespace std;

class Solution {
private:
    char solveOperator(char &op, vector<char>& chStore) {
        if (op == '!') return chStore[0] == 't' ? 'f' : 't';

        if (op == '|') {
            // for (char& ch : chStore) {
            //     if (ch == 't') return 't';
            // }
            // return 'f';

            return any_of(begin(chStore), end(chStore), [](char ch) { return ch == 't'; }) ? 't' : 'f'; // one liner
        }

        // else op == '&'
        // for (char& ch : chStore) {
        //     if (ch == 'f') return 'f';
        // }
        // return 't';

        return any_of(chStore.begin(), chStore.end(), [](char ch) { return ch == 'f'; }) ? 'f' : 't';
    }
public:
    // TC : O(N)
    // SC : O(N) Where N = expression length
    bool parseBoolExpr(string expression) {
        stack<int> stk;
        for (char &ch : expression) {
            if (ch == ',') continue;
            if (ch == ')') {
                vector<char> charStore;
                while (stk.top() != '(') {
                    charStore.push_back(stk.top());
                    stk.pop();
                }

                stk.pop(); // '('
                
                // Operators : '!', '|', '&'
                char op = stk.top();
                stk.pop();

                stk.push(solveOperator(op, charStore));
            } else {
                stk.push(ch);
            }
        }
        return stk.top() != 'f';
    }
};

File: Leetcode Daily Challenge/016-October-2024/020-parsing-a-boolean-expression.java
// https://leetcode.com/problems/parsing-a-boolean-expression/


import java.util.*;

class Solution {
    public boolean parseBoolExpr(String expression) {
        Stack<Character> stk = new Stack<>();

        for (char ch : expression.toCharArray()) {
            if (ch == ',') continue;

            if (ch == ')') {
                ArrayList<Character> vec = new ArrayList<>();

                while (stk.peek() != '(') {
                    char topCh = stk.peek();
                    stk.pop();
                    vec.add(topCh);
                }

                stk.pop();  // '('
                
                char op = stk.peek();    // '!', '|', '&'
                stk.pop();
                
                char getAnswer = solveOperator(vec, op);
                stk.push(getAnswer);
            } else {
                stk.push(ch);
            }
        }
        return stk.peek() != 'f';
    }

    private char solveOperator(ArrayList<Character> arr, char op) {
        if (op == '!') {
            return arr.get(0) == 'f' ? 't' : 'f';
        }

        if (op == '&') {
            for (char ch : arr) {
                if (ch == 'f') return 'f';
            }
            return 't';
        }

        // '|'
        for (char ch : arr) {
            if (ch == 't') return 't';
        }
        return 'f';
    }
}